File: Section6_Mapping_with_known_Poses/bumperbot_ws/src/bumperbot_py_examples/bumperbot_py_examples/simple_qos_publisher.py
import rclpy
from rclpy.node import Node
from std_msgs.msg import String
from rclpy.qos import QoSProfile, QoSDurabilityPolicy, QoSReliabilityPolicy

class SimpleQosPublisher(Node):

    def __init__(self):
        super().__init__('simple_qos_publisher')
        self.qos_profile_pub = QoSProfile(depth=10)
        self.counter = 0

        self.declare_parameter('reliability', 'system_default')
        self.declare_parameter('durability', 'system_default')

        reliability = self.get_parameter('reliability').get_parameter_value().string_value
        durability = self.get_parameter('durability').get_parameter_value().string_value

        if reliability == 'best_effort':
            self.qos_profile_pub.reliability = QoSReliabilityPolicy.BEST_EFFORT
            self.get_logger().info('[Reliability] : Best Effort')
        elif reliability == 'reliable':
            self.qos_profile_pub.reliability = QoSReliabilityPolicy.RELIABLE
            self.get_logger().info('[Reliability] : Reliable')
        elif reliability == 'system_default':
            self.qos_profile_pub.reliability = QoSReliabilityPolicy.SYSTEM_DEFAULT
            self.get_logger().info('[Reliability] : System Default')
        else:
            self.get_logger().error('Selected Reliability QoS: %s doesn\'t exist!' % reliability)
            return

        if durability == 'volatile':
            self.qos_profile_pub.durability = QoSDurabilityPolicy.VOLATILE
            self.get_logger().info('[Durability] : Volatile')
        elif durability == 'transient_local':
            self.qos_profile_pub.durability = QoSDurabilityPolicy.TRANSIENT_LOCAL
            self.get_logger().info('[Durability] : Transient Local')
        elif durability == 'system_default':
            self.qos_profile_pub.durability = QoSDurabilityPolicy.SYSTEM_DEFAULT
            self.get_logger().info('[Durability] : System Default')
        else:
            self.get_logger().error('Selected Durability QoS: %s doesn\'t exist!' % durability)
            return

        self.pub = self.create_publisher(String, 'chatter', self.qos_profile_pub)
        self.timer = self.create_timer(1.0, self.timer_callback)
        self.get_logger().info('Publishing at 1 Hz')

    def timer_callback(self):
        message = String()
        message.data = 'Hello ROS 2 QoS - counter: %d' % self.counter
        self.pub.publish(message)
        self.counter += 1

def main(args=None):
    rclpy.init(args=args)
    node = SimpleQosPublisher()
    rclpy.spin(node)
    node.destroy_node()
    rclpy.shutdown()

if __name__ == '__main__':
    main()

File: Section10_SLAM/bumperbot_ws/src/bumperbot_cpp_examples/src/simple_qos_publisher.cpp
#include <chrono>
#include <string>
#include <memory>

#include "rclcpp/rclcpp.hpp"
#include "std_msgs/msg/string.hpp"

using namespace std::chrono_literals;

class SimpleQosPublisher : public rclcpp::Node
{
public:
  SimpleQosPublisher() : Node("simple_qos_publisher"), qos_profile_pub_(10), counter_(0)
  {
    declare_parameter<std::string>("reliability", "system_default");
    declare_parameter<std::string>("durability", "system_default");

    const auto reliability = get_parameter("reliability").as_string();
    const auto durability = get_parameter("durability").as_string();

    if(reliability == "best_effort")
    {
      qos_profile_pub_.reliability(RMW_QOS_POLICY_RELIABILITY_BEST_EFFORT);
      RCLCPP_INFO(get_logger(), "[Reliability] : Best Effort");
    }
    else if(reliability == "reliable")
    {
      qos_profile_pub_.reliability(RMW_QOS_POLICY_RELIABILITY_RELIABLE);
      RCLCPP_INFO(get_logger(), "[Reliability] : Reliable");
    }
    else if(reliability == "system_default")
    {
      qos_profile_pub_.reliability(RMW_QOS_POLICY_RELIABILITY_SYSTEM_DEFAULT);
      RCLCPP_INFO(get_logger(), "[Reliability] : System Default");
    }
    else
    {
      RCLCPP_ERROR_STREAM(get_logger(), "Selected Reliability QoS: " << reliability << " doesn't exists!");
      return;
    }

    if(durability == "volatile")
    {
      qos_profile_pub_.durability(RMW_QOS_POLICY_DURABILITY_VOLATILE);
      RCLCPP_INFO(get_logger(), "[Durability] : Volatile");
    }
    else if(durability == "transient_local")
    {
      qos_profile_pub_.durability(RMW_QOS_POLICY_DURABILITY_TRANSIENT_LOCAL);
      RCLCPP_INFO(get_logger(), "[Durability] : Transient Local");
    }
    else if(durability == "system_default")
    {
      qos_profile_pub_.durability(RMW_QOS_POLICY_DURABILITY_SYSTEM_DEFAULT);
      RCLCPP_INFO(get_logger(), "[Durability] : System Default");
    }
    else
    {
      RCLCPP_ERROR_STREAM(get_logger(), "Selected Durability QoS: " << durability << " doesn't exists!");
      return;
    }

    pub_ = create_publisher<std_msgs::msg::String>("chatter", qos_profile_pub_);
    timer_ = create_wall_timer(1s, std::bind(&SimpleQosPublisher::timerCallback, this));
    RCLCPP_INFO(get_logger(), "Publishing at 1 Hz");
  }

  void timerCallback()
  {
    auto message = std_msgs::msg::String();
    message.data = "Hello ROS 2 QoS - counter:" + std::to_string(counter_++);
    pub_->publish(message);
  }

private:
  rclcpp::Publisher<std_msgs::msg::String>::SharedPtr pub_;
  rclcpp::QoS qos_profile_pub_;
  rclcpp::TimerBase::SharedPtr timer_;
  unsigned int counter_;
};

int main(int argc, char *argv[])
{
  rclcpp::init(argc, argv);
  auto node = std::make_shared<SimpleQosPublisher>();
  rclcpp::spin(node);
  rclcpp::shutdown();
  return 0;
}